packages/frameworks-qwik/src/index.ts

```diff
@@ -0,0 +1,382 @@
+          res.headers.delete("set-cookie")
+          fixCookies(req)
+        }
+        throw req.send(res)
+      } else {
+        const { data, cookie } = await getSessionData(req, patchedAuthOptions)
+        req.sharedMap.set("session", data)
+        if (cookie) {
+          req.headers.set("set-cookie", cookie)
+          fixCookies(req)
+        }
+      }
+    }
+  }
+
+  return {
+    useAuthSignIn,
+    useAuthSignOut,
+    useAuthSession,
+    onRequest,
+  }
+}
+
+export const qwikAuth$ = /*#__PURE__*/ implicit$FirstArg(qwikAuthQrl)
+
+async function authAction(
+  body: URLSearchParams | undefined,
+  req: RequestEventCommon,
+  path: string,
+  authOptions: QwikAuthConfig
+) {
+  const request = new Request(new URL(path, req.request.url), {
+    method: req.request.method,
+    headers: req.request.headers,
+    body: body,
+  })
+  request.headers.set("content-type", "application/x-www-form-urlencoded")
+  const res = (await Auth(request, {
+    ...authOptions,
+    skipCSRFCheck,
+  })) as Response
+
+  const cookies: string[] = []
+  res.headers.forEach((value, key) => {
+    if (key === "set-cookie") {
+      // while browsers would support setting multiple cookies, the fetch implementation does not, so we join them later.
+      cookies.push(value)
+    } else if (!req.headers.has(key)) {
+      req.headers.set(key, value)
+    }
+  })
+
+  if (cookies.length > 0) {
+    req.headers.set("set-cookie", cookies.join(", "))
+  }
+
+  fixCookies(req)
+
+  try {
+    return await res.json()
+  } catch (error) {
+    return await res.text()
+  }
+}
+
+const fixCookies = (req: RequestEventCommon) => {
+  req.headers.set("set-cookie", req.headers.get("set-cookie") || "")
+  const cookie = req.headers.get("set-cookie")
+  if (cookie) {
+    req.headers.delete("set-cookie")
+    splitCookiesString(cookie).forEach((cookie) => {
+      const { name, value, ...rest } = parseString(cookie)
+      req.cookie.set(name, value, rest as any)
+    })
+  }
+}
+
+export const ensureAuthMiddleware = (req: RequestEventCommon) => {
+  const isLoggedIn = req.sharedMap.has("session")
+  if (!isLoggedIn) {
+    throw req.error(403, "sfs")
+  }
+}
+
+const defaultCallbackURL = (req: RequestEventCommon) => {
+  req.url.searchParams.delete("qaction")
+  return req.url.href
+}
+
+async function getSessionData(
+  req: RequestEventCommon,
+  patchedAuthOptions: QwikAuthConfig
+): GetSessionResult {
+  const url = new URL("/api/auth/session", req.url)
+  const response = (await Auth(
+    new Request(url, { headers: req.headers }),
+    patchedAuthOptions
+  )) as Response
+
+  const { status = 200 } = response
+
+  const data = await response.json()
+  const cookie = response.headers.get("set-cookie")
+  if (!data || !Object.keys(data).length) {
+    return { data: null, cookie }
+  }
+  if (status === 200) {
+    return { data, cookie }
+  }
+
+  throw new Error(data.message)
+}
+
+const patchAuthOptions = async (
```

Comment:
nit: let's remove this function, I think it will be simpler to just call the two lines for each actions